### PR TITLE
Visit Stmt children for null deref check.

### PR DESCRIPTION
**Pull request by CristinaCristescu**
The previous version of the Null deref visitor was defaulting to
VisitStmt when the Stmt type was unknown, thus allowing for some
types of Stmt X for which a VisitX was not implemented to default
to VisitStmt and not check the content of it. In order to check
any type of Stmt for possible subexpressions which contain s Stmt
that derefrences a Null, the VisitStmt was extended to check its
children.

Only the Stmt types, X, which directly containing a null deref expression
have an explicit VisitX with the exception of CXXMemberCallExpr,
because this type would otherwise default to VisitCallExpr which has
a different check for its structure.

### CODE REVIEW

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 1]
```diff
@@ -107,6 +107,11 @@ namespace cling {
     }
 
     NodeContext VisitStmt(Stmt* S) {
+      for (auto I = S->child_begin(), E = S->child_end(); I != E; ++I) {
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: or
```
for (auto S: S->children()) {
  if (S)
    Visit(S)
}
```

Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Do we care about the return value of `Visit()`?

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 2]
```diff
@@ -107,6 +107,11 @@ namespace cling {
...
+        if (*I)
+          NodeContext nc = Visit(*I);
+      }
+
       return NodeContext(S);
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Given that we now always return an unchanged `NodeContext` - can't we just return nothing? Or do we ever care about the `NodeContext` that `Visit()` returns?

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 3]
```diff
@@ -122,79 +127,42 @@ namespace cling {
...
-      return result;
-    }
-
-    NodeContext VisitCastExpr(CastExpr* CE) {
-      NodeContext castExpr = Visit(CE->getSubExpr());
-      if (castExpr.isSingleStmt())
-        CE->setSubExpr((clang::Expr*)castExpr.getStmt());
+      if (!stmts.empty()) {
+        llvm::ArrayRef<Stmt*> stmtsRef(stmts.data(), stmts.size());
+        CS->setStmts(C, const_cast<Stmt**>(&stmtsRef.front()), stmts.size());
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: According to the doc, `setStmts` takes an `ArrayRef` (in the clang master...). Can we just pass in `stmtsRef` itself, or do we really need to pass in the `Stmt**`?

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 4]
```diff
@@ -107,6 +107,11 @@ namespace cling {
     }
 
     NodeContext VisitStmt(Stmt* S) {
+      for (auto I = S->child_begin(), E = S->child_end(); I != E; ++I) {
+        if (*I)
+          NodeContext nc = Visit(*I);
+      }
+
       return NodeContext(S);
     }
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Do we really ever need to update the address of the elements inside a `CompoundStmt`? I don't think so... Then we don't need `VisitCompoundStmt()` at all...

Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: ...and then `Visit()` can return `void` (see the template arg of the `RecursiveXYZVisitor`

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 5]
```diff
@@ -122,79 +127,42 @@ namespace cling {
...
-      if (rhs.isSingleStmt()) {
-        // FIXME:we need to loop from 0 to n-1
-        BinOp->setRHS((clang::Expr*)rhs.getStmt());
-      }
-      if (lhs.isSingleStmt()) {
-        // FIXME:we need to loop from 0 to n-1
-        BinOp->setLHS((clang::Expr*)lhs.getStmt());
+        NodeContext newSubExpr = SynthesizeCheck(UnOp->getLocStart(),
+                                 UnOp->getSubExpr());
+        if (newSubExpr.isSingleStmt()) {
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Don't we know exactly what we synthesize?

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 6]
```diff
@@ -122,79 +127,42 @@ namespace cling {
...
+          assert ((clang::Expr*)newSubExpr.getStmt()
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Did you mean to say `llvm::isa<clang::Expr>(newSubExpr.getStmt())` here? See http://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 7]
```diff
@@ -122,79 +127,42 @@ namespace cling {
...
+                  && "Visiting Stmt which is not an Expr!");
+          UnOp->setSubExpr((clang::Expr*)newSubExpr.getStmt());
+        }
       }
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Isn't there a missing `else`, where we need to continue descending and visiting the children? Actually - that's not even an `else`, we should just always to that. So you might as well call `VisitStmt(UnOp)` from here.

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 8]
```diff
@@ -122,79 +127,42 @@ namespace cling {
...
         NodeContext newBase = SynthesizeCheck(ME->getLocStart(), ME->getBase());
         if (newBase.isSingleStmt())
+          assert ((clang::Expr*)newBase.getStmt()
+                  && "Visiting Stmt which is not an Expr!");
           ME->setBase((clang::Expr*)newBase.getStmt());
       }
 
-      NodeContext result(ME);
-      return result;
+      return NodeContext(ME);
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: See above, we might want to recurse into `VisitStmt(ME)` here. Not sure whether we should do that before the call to `SynthesizeCheck()` or after.

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 9]
```diff
@@ -122,79 +127,42 @@ namespace cling {
...
-      NodeContext result(BinOp);
-      return result;
+      return NodeContext(UnOp);
     }
 
     NodeContext VisitMemberExpr(MemberExpr* ME) {
       if (ME->isArrow()) {
         NodeContext newBase = SynthesizeCheck(ME->getLocStart(), ME->getBase());
         if (newBase.isSingleStmt())
+          assert ((clang::Expr*)newBase.getStmt()
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: See `isa` comment above.

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 10]
```diff
@@ -203,24 +171,31 @@ namespace cling {
           if (ArgIndexs.test(index)) {
             // Get the argument with the nonnull attribute.
             Expr* Arg = CE->getArg(index);
+            NodeContext newArg = SynthesizeCheck(Arg->getLocStart(), Arg);
+            if (newArg.isSingleStmt()) {
+              assert ((clang::Expr*)newArg.getStmt()
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: See `isa` comment above. Actually - given that this shows up three times - can we just let `SynthesizeCheck` return an `Expr*`?

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 11]
```diff
@@ -203,24 +171,31 @@ namespace cling {
...
             CE->setArg(index,
-                      (clang::Expr*)SynthesizeCheck(Arg->getLocStart(), Arg));
+                      (clang::Expr*)newArg.getStmt());
+            }
           }
         }
       }
-      return result;
+
+      return NodeContext(CE);
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: See `recursion` note above.

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 12]
```diff
@@ -203,24 +171,31 @@ namespace cling {
...
+        NodeContext newME = Visit(Callee);
+        if (newME.isSingleStmt())
+          assert ((clang::Expr*)newME.getStmt()
+                      && "Visiting Stmt which is not an Expr!");
+          CME->setCallee((clang::Expr*)newME.getStmt());
       }
 
-      NodeContext result(CME);
-      return result;
+      return NodeContext(CME);
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Do we still need this function? I don't see it calling `SynthesizeCheck()` - so it should be covered by `VisitStmt()`, right?

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 1 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Done, void!

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 4 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Removed

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 3 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Irrelevant now

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 5 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Yes, changed to Expr*

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 6 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Not needed anymore

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 7 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Added the Visit(UnOp->getSubExpr())

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 9 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Done

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 8 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Added before SynthesizedCheck()

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 10 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Always wanted to change that! :-)

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 11 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Seen

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 12 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Covered by adding Visit in the VisitCallExpr()

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 13]
```diff
@@ -33,58 +33,7 @@ namespace cling {
...
-      m_Stmts.insert(m_Stmts.begin(), S);
-    }
-
-    void append(clang::Stmt* S) {
-      m_Stmts.push_back(S);
-    }
-  };
-
-  class IfStmtInjector : public StmtVisitor<IfStmtInjector, NodeContext> {
+  class IfStmtInjector : public StmtVisitor<IfStmtInjector, void> {
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: While we're waiting for roottest :-) Can we rename this to something else? E.g. PointerCheckInjector?

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 14]
```diff
@@ -101,100 +50,37 @@ namespace cling {
   public:
     IfStmtInjector(Sema& S) : m_Sema(S), m_Context(S.getASTContext()),
     m_LookupResult(0) {}
-    CompoundStmt* Inject(CompoundStmt* CS) {
-      NodeContext result = VisitCompoundStmt(CS);
-      return cast<CompoundStmt>(result.getStmt());
+    void Inject(CompoundStmt* CS) {
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Can we just call `Visit(CS)` from the `NullDerefTransformer` and get rid of `Inject()`?

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 15]
```diff
@@ -101,100 +50,37 @@ namespace cling {
...
-      // Here we might get if(check) throw; binop rhs.
-      NodeContext rhs = Visit(BinOp->getRHS());
-      // Here we might get if(check) throw; binop lhs.
-      NodeContext lhs = Visit(BinOp->getLHS());
-
-      if (rhs.isSingleStmt()) {
-        // FIXME:we need to loop from 0 to n-1
-        BinOp->setRHS((clang::Expr*)rhs.getStmt());
+        Expr* newSubExpr = SynthesizeCheck(UnOp->getLocStart(),
+                                 UnOp->getSubExpr());
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Indentation seems to be off? Can you align the `UnOp->`s (i.e. func args should be aligned across lines, unless the line is longer than 80 chars)

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 16]
```diff
@@ -203,28 +89,15 @@ namespace cling {
...
-          CME->setCallee((clang::Expr*)ME.getStmt());
-      }
-
-      NodeContext result(CME);
-      return result;
     }
 
   private:
-    Stmt* SynthesizeCheck(SourceLocation Loc, Expr* Arg) {
+    Expr* SynthesizeCheck(SourceLocation Loc, Expr* Arg) {
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: `Loc` seems to always be `Arg->getLocStart()`. Do we really need to pass it in?

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 17]
```diff
@@ -203,28 +89,15 @@ namespace cling {
           if (ArgIndexs.test(index)) {
             // Get the argument with the nonnull attribute.
             Expr* Arg = CE->getArg(index);
-            CE->setArg(index,
-                      (clang::Expr*)SynthesizeCheck(Arg->getLocStart(), Arg));
+            Expr* newArg = SynthesizeCheck(Arg->getLocStart(), Arg);
+            CE->setArg(index, newArg);
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Once the call to `Arg->getLocStart()` is gone you could even combine these into one or two statements.

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 13 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Done!

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 15 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: It does not fit.. :-( unless I rename it?

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 17 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Ok

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 16 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Ok, that is true! And it reduces the code quite a bit!

**Comment by karies** [comment]
Thanks, merged! And congrats - you have removed half of NullDerefProtectionTransformer.cpp :-)

**Pull request closed without merging**